### PR TITLE
Add Inferentai2 and Optimum Neuron Support

### PR DESCRIPTION
**Pull request by philschmid**
# What does this PR do? 

This PR add support for infernetia2. To deploy a model on Inferentia2 you have 3 options:
* Provide an already compiled model with a `model.neuron` file as `HF_MODEL_ID`, .e.g. `optimum/tiny_random_bert_neuron`
* Provide the `HF_OPTIMUM_BATCH_SIZE` and `HF_OPTIMUM_SEQUENCE_LENGTH` environment variables to compile the model on the fly, e.g. `HF_OPTIMUM_BATCH_SIZE=1 HF_OPTIMUM_SEQUENCE_LENGTH=128`
* Include `neuron` dictionary in the [config.json](https://huggingface.co/optimum/tiny_random_bert_neuron/blob/main/config.json) file in the model archive, e.g. `neuron: {"static_batch_size": 1, "static_sequence_length": 128}`

The currently supported tasks can be found [here](https://huggingface.co/docs/optimum-neuron/en/package_reference/supported_models). If you plan to deploy an LLM, we recommend taking a look at [Neuronx TGI](https://huggingface.co/blog/text-generation-inference-on-inferentia2), which is purposly build for LLMs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -157,7 +157,14 @@ The custom module can override the following methods:
 * `output_fn(prediction, accept)`: overrides the default method for postprocessing, the return value `result` will be the respond of your request(e.g.`JSON`). The inputs are `predictions`, the result of the `predict()` method and `accept` the return accept type from the HTTP Request, e.g. `application/json`
 
 
+## 🏎️ Deploy Models on AWS Inferentia2
 
+The SageMaker Hugging Face Inference Toolkit provides support for deploying Hugging Face on AWS Inferentia2. To deploy a model on Inferentia2 you have 3 options:
+* Provide an already compiled model with a `model.neuron` file as `HF_MODEL_ID`, .e.g. `optimum/tiny_random_bert_neuron`
```
Comment (JingyaHuang) [inline README.md:163]:
```suggestion
* Provide `HF_MODEL_ID`, the model repo id on huggingface.co which contains the compiled model under `.neuron` format. e.g. `optimum/bge-base-en-v1.5-neuronx`
```

File: src/sagemaker_huggingface_inference_toolkit/optimum_utils.py [hunk 2]
```diff
@@ -38,21 +38,25 @@ def get_input_shapes(model_dir):
...
-                f"Input shapes found in config file. Using input shapes from config with batch size {input_shapes['batch_size']} and sequence length {input_shapes['sequence_length']}"
-            )
-            if os.environ.get("HF_OPTIMUM_BATCH_SIZE", None) is not None:
-                logger.warning(
-                    "HF_OPTIMUM_BATCH_SIZE environment variable is set. Environment variable will be ignored and input shapes from config file will be used."
-                )
-            if os.environ.get("HF_OPTIMUM_SEQUENCE_LENGTH", None) is not None:
-                logger.warning(
-                    "HF_OPTIMUM_SEQUENCE_LENGTH environment variable is set. Environment variable will be ignored and input shapes from config file will be used."
+        if hasattr(config, "neuron"):
```
Comment (JingyaHuang) [inline src/sagemaker_huggingface_inference_toolkit/optimum_utils.py:41]: And what if the neuron section doesn't exist in the config? Shall we raise or warn here?

File: src/sagemaker_huggingface_inference_toolkit/transformers_utils.py [hunk 3]
```diff
@@ -202,7 +207,9 @@ def _load_model_from_hub(
     # check if safetensors weights are available
     if framework == "pytorch":
         files = HfApi().model_info(model_id).siblings
-        if any(f.rfilename.endswith("safetensors") for f in files):
+        if is_optimum_neuron_available() and any(f.rfilename.endswith("neuron") for f in files):
+            framework = "neuronx"
```
Comment (JingyaHuang) [inline src/sagemaker_huggingface_inference_toolkit/transformers_utils.py:211]: In optimum, both models compiled for inf1 and inf2 are in `.neuron` format. If we want to distinguish that, we could leverage the compiler type under `config.neuron`: `neuronx-cc` for neuronx and `neuron-cc` for neuron.

**Review by JingyaHuang — APPROVED** [review]
LGTM, thanks @philschmid. Just left some small nits.

File: src/sagemaker_huggingface_inference_toolkit/optimum_utils.py (hunk 2 above)
Comment (philschmid) [reply src/sagemaker_huggingface_inference_toolkit/optimum_utils.py:41]: if it doesn't exist the env vars will be checked. Its just that the priority is config > env > error

File: src/sagemaker_huggingface_inference_toolkit/transformers_utils.py (hunk 3 above)
Comment (philschmid) [reply src/sagemaker_huggingface_inference_toolkit/transformers_utils.py:211]: thats fine since inf1 should be supported for encoder tasks as well.

**Pull request merged**